src/client/src/producer.rs

```diff
@@ -93,16 +110,18 @@ impl TopicProducer {
 
         debug!("connect to replica leader at: {}", spu_client);
 
-        send_record_raw(spu_client, &replica, None, record).await
+        let record = Bytes::from(Vec::from(buffer.as_ref()));
+        let records = &[(None, record)];
+        send_records_raw(spu_client, &replica, records).await?;
+        Ok(())
     }
 }
 
 /// Sends record to a target server (Kf, SPU, or SC)
-async fn send_record_raw<F: SerialFrame>(
+async fn send_records_raw<F: SerialFrame>(
     mut leader: F,
     replica: &ReplicaKey,
-    key: Option<&[u8]>,
-    value: &[u8],
+    records: &[(Option<Bytes>, Bytes)],
```

Comment:
This could be Vec<DefaultRecord>, that way, can eliminate unnecessary conversion

Reply:
Which unnecessary conversion are you referring to? The reason to use a slice rather than an array is that you can pass a slice of any size, whereas the array must be a fixed size. I also don't think a Vec is necessary here, unless you are talking about using Vec instead of Bytes? But if that's what you're asking I'm not sure what the advantage to that is either.

Reply:
This is only used internally.  `send_all` uses Vec.  So you are reconstructing Vec twice rather than just constructing them once.

Reply:
If you look at how the `records: &[(Option<Bytes>, Bytes)]` slice gets used inside of `send_records_raw`, it is simply being iterated over and the inner `Bytes` objects are being loaded into the request. I don't think there would be any advantage to passing the entire owned Vec through this call, there are no extra allocations being made.

In fact, if you look at the other usage of `send_records_raw` in the `send_record` method, it is actually directly constructing a slice on the stack, and does not require the use of Vec at all. If we were to change `send_records_raw` to require a Vec, then `send_record` would gain an extra unnecessary allocation per record sent.

Reply:
Updated based on our conversation earlier

Reply:
Looks good